Implement redirect reaction for Metabot

retro opened the pull request:
Redirect reaction redirects the user to the results page generated by a tool.
Currently `query-metric` and `generate-insights` tools use this reaction.

This also introduces a concept of "terminating" reactions. When a tool returns a
terminating reaction, this tool call result is not sent back to the LLM, instead
it returns the envelope and the frontend handles any sideffects.

retro committed:
Implement redirect reaction for Metabot

Redirect reaction redirects the user to the results page generated by a tool.
Currently `query-metric` and `generate-insights` tools use this reaction.

This also introduces a concept of "terminating" reactions. When a tool returns a
terminating reaction, this tool call result is not sent back to the LLM, instead
it returns the envelope and the frontend handles any sideffects.

metamben commented on enterprise/backend/src/metabase_enterprise/metabot_v3/tools/generate_insights.clj:
```diff
@@ -23,4 +23,5 @@
                     id)]
-    {:output (str (public-settings/site-url) "/auto/dashboard/" entity-type "/" entity-id)}))
+    {:output (str (public-settings/site-url) "/auto/dashboard/" entity-type "/" entity-id)
+     :reactions [{:type :metabot.reaction/redirect :url (str "/auto/dashboard/" entity-type "/" entity-id)}]}))
```
nit: I suggest extracting and re-using `(str "/auto/dashboard/" entity-type "/" entity-id)`.

metamben commented on enterprise/backend/src/metabase_enterprise/metabot_v3/tools/show_results_to_user.clj:
```diff
@@ -16,5 +16,6 @@
                          (.getBytes "UTF-8")
                          codecs/bytes->b64-str)]
-      {:output (str "Results can be seen at: " (public-settings/site-url) "/question#" query-hash)})
+      {:output (str "Results can be seen at: " (public-settings/site-url) "/question#" query-hash)
+       :reactions [{:type :metabot.reaction/redirect :url (str "/question#" query-hash)}]})
```
nit: I suggest extracting and re-using `(str "/question#" query-hash)`.

retro committed:
Stylistic improvements

- Implement `terminating-reaction?` function and implement a cleaner way of determining
if reactions vector contains a terminating reaction
- Extract results url for the `show-results-to-user` and `generate-insights` tools and reuse it
for the message body and the redirect reaction

metamben commented on enterprise/backend/src/metabase_enterprise/metabot_v3/reactions.clj:
```diff
@@ -90,8 +89,12 @@
   [reactions]
-  (let [reaction-types (->> reactions (map :type) set)]
-    (seq (set/intersection terminating-reaction-types reaction-types))))
+  (some terminating-reaction? reactions))
```
I would have made `terminating-reaction?` public and inline `has-terminating-reaction?`.
